File: chap-3/cpp/ds/queue.cpp
#include "queue.hpp"

template <class T>
void Queue<T>::enqueue(T data){
	if(this->last){
		Node<T> *newNode = new Node<T>(data);
		this->last->next = newNode;
		this->last = newNode;
	}
}

template <class T>
Node<T> Queue<T>::dequeue(){
	if(this->first){
		Node<T> result(this->first->data);
		Node<T> *toDelete = this->first;
		this->first = this->first->next;
		delete toDelete;
		return result;

	}
	return 0;
}

template <class A>
std::ostream & operator <<(std::ostream &out, const Queue<A> &q){
	
	if(!q.first){
		out << "Queue Empty" << std::endl;
	}else{
		out << "Queue" << std::endl;
		Node<A> *tmp= q.first;
		out << "First" << std::endl;
		out << "=====" << std::endl;
		while(tmp){
			out << tmp->data << std::endl;
			tmp = tmp->next;
		}
		out << "=====" << std::endl;
		out << "Last" << std::endl;
	}
	return out;
}

File: chap-3/python/queue.py
"""
    Nii Mante
    Queue implementation 
"""
from node import Node

class Queue:
    def __init__(self, data = None):
        if data:
            self.first = Node(data)
            self.last = self.first
        else:
            self.first = None
            self.last = None

    def __str__(self):
        tmp = self.first
        out_str = "Queue\n"
        out_str = out_str + "====\n"
        while tmp:
            out_str = out_str + str(tmp.data) + "\n"
            tmp = tmp.next_node

        out_str = out_str + "\n====\n"
        return out_str
            

    def enqueue(self, data):
        """
            Push an item to the end of the queue
        """
        if self.last:
            self.last.next_node = Node(data)
        else:
            self.last = Node(data)

    def dequeue(self):
        """
            Pop an item from the beginning of the queue
        """
        result_node = Node(self.first.data) 
        self.first = self.first.next_node
        return result_node
